include/picongpu/algorithms/AssignedTrilinearInterpolation.hpp

```diff
@@ -52,64 +52,75 @@ namespace picongpu
         /** Does a 3D trilinear field-to-point interpolation for
          * arbitrary assignment function and arbitrary field_value types.
          *
-         * @tparam T_AssignmentFunction function for assignment
          * @tparam T_begin lower margin for interpolation
          * @tparam T_end upper margin for interpolation
+         * @tparam T_AssignmentFunction type of the array with shape functors
          *
          * @param cursor cursor pointing to the field
-         * @param pos position of the interpolation point
+         * @param shapeFunctors Array with d shape functors, where d is the dimensionality of the field represented by
+         *                      cursor. The shape functor must have the interface to call
+         *                      `operator()(relative_grid_point)` and return the assignment value for the given grid
+         *                      point.
          * @return sum over: field_value * assignment
          *
          * interpolate on grid points in range [T_begin;T_end]
          */
-        template<typename T_AssignmentFunction, int T_begin, int T_end, typename T_Cursor>
-        HDINLINE static auto interpolate(const T_Cursor& cursor, const float3_X& pos) ->
-            typename ::pmacc::result_of::Functor<AssignedTrilinearInterpolation, T_Cursor>::type
+        template<int T_begin, int T_end, typename T_Cursor, typename T_AssignmentFunction>
+        HDINLINE static auto interpolate(
+            const T_Cursor& cursor,
+            const pmacc::memory::Array<T_AssignmentFunction, 3>& shapeFunctors)
         {
-            using type = typename ::pmacc::result_of::Functor<AssignedTrilinearInterpolation, T_Cursor>::type;
-
             [[maybe_unused]] constexpr auto iterations = T_end - T_begin + 1;
-            auto result_z = type(0.0);
+
+            /* The implementation assumes that x is the fastest moving index to iterate over contiguous memory
+             * e.g. a row, to optimize memory fetch operations.
+             */
+            auto result_z = float_X(0.0);
             PMACC_UNROLL(iterations)
             for(int z = T_begin; z <= T_end; ++z)
             {
-                auto result_y = type(0.0);
+                auto result_y = float_X(0.0);
                 PMACC_UNROLL(iterations)
                 for(int y = T_begin; y <= T_end; ++y)
                 {
-                    auto result_x = type(0.0);
+                    auto result_x = float_X(0.0);
```

Comment:
I think it is not right that we hardcode `float_X` here, as in principle these utilities are supposed to work for any PIConGPU field, not just `E` or `B`. And i think this is the reason we do not assume `d == simDim` here but work for both `d = 2` and `d = 3`. So imo we should keep the `type` alias for element type as before, just it was also not correctly assigned previously. It should be `using type = decltype(*cursor(0, 0, 0));`